File: h3controllers/readOnlyAfterController.py
# Copyright [2019] [FORTH-ICS]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time 
import argparse
import struct

import pyh3lib

def ReadOnlyAfter(h3):
    """
    Set's the permissions to read only in all objects that have the ReadOnlyAfter attribute and the "now" time exceeds
    the time that is specified in the ReadOnlyAfter plus the time from the last time that the object has been modified.
    
    :param now: the time, now
    :type now: float
    :returns: nothing
    """

    now = time.clock_gettime(time.CLOCK_REALTIME)

    # list all the buckets 
    for h3_bucket in h3.list_buckets():
        done   = False
        offset = 0
        
        # list all the objects that have the ExpiresAt attribute
        while not done:
            h3_objects = h3.list_objects_with_metadata(h3_bucket, "ReadOnlyAfter", offset)

            for h3_object in h3_objects:
                # the h3_object contains the object's name
                read_only_after = h3.read_object_metadata(h3_bucket, h3_object, "ReadOnlyAfter")

                if read_only_after != b'':
                    # try to parse
                    try:
                        h3_object_remove_timestamp = struct.unpack('d', read_only_after)
                    except struct.error:
                        continue

                    h3_object_info = h3.info_object(h3_bucket, h3_object)

                    # Check if we must change the permissions of the object to read only
                    if ((not h3_object_info.read_only) and (now >= (h3_object_info.last_modification + h3_object_remove_timestamp[0]))):
                        h3.make_object_read_only(h3_bucket, h3_object)

            done    = h3_objects.done
            offset  = h3_objects.nextOffset

def main(cmd=None):
    parser = argparse.ArgumentParser(description='ReadOnlyAfter Controller')
    parser.add_argument('--storage', required=True, help=f'H3 storage URI')
    
    args = parser.parse_args(cmd)
    config_path = args.storage 
    
    if config_path:
        h3 = pyh3lib.H3(config_path)

        ReadOnlyAfter(h3)
    else:
        parser.print_help(sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()
         

File: h3controllers/ch3controllers/readOnlyAfterController.c
// Copyright [2019] [FORTH-ICS]
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <time.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <h3lib/h3lib.h>

#define TIMESPEC_TO_DOUBLE(t)   (t.tv_sec + ((double)t.tv_nsec / 1000000000ULL))
#define TRUE                    1

void PrintUsage() {
    fprintf(stderr, "Usage:\n");
    fprintf(stderr, "\t-s <string>  H3 storage URI\n");
    fprintf(stderr, "\t-h           This help messsage\n\n");
}

int main(int argc, char* argv[]) {    
    char* url;

    if (argc < 2) {
        PrintUsage();
        return 1;
    }

    int opt;
    while ((opt = getopt(argc, argv, ":s:h")) != -1) {
        switch (opt) {
            case 's':{
                url = strdup(optarg);
            }break;
            
            case 'h':{
                PrintUsage();
                
                return 0;
            }break;
        
            default:{
                PrintUsage();
                
                return 1;
            }
        }
    }  
    
    // Get the time now
    struct timespec now;
    clock_gettime(CLOCK_REALTIME, &now);

    // Authentication
    H3_Auth auth;
    auth.userId = 0;
    
    H3_Handle h3_handle = H3_Init(url);
    if (!h3_handle) {
        fprintf(stderr, "[ERROR] : fail to init H3.\n");
        return 1;
    }

    H3_Name buckets = NULL;
    uint32_t totalBuckets; 

    // List all the buckets    
    if (H3_ListBuckets(h3_handle, &auth, &buckets, &totalBuckets) == H3_SUCCESS) {
        H3_Name current_bucket;
        uint32_t current_bucket_pos = 0;
        size_t current_bucket_len = 0;
        int bucket;
        for (bucket = 0; bucket < totalBuckets; bucket++) {
            current_bucket = &(buckets[current_bucket_pos]);
            current_bucket_len = strlen(current_bucket);
            current_bucket_pos += current_bucket_len;
            while (buckets[current_bucket_pos] == '\0')
                current_bucket_pos++;
            
            H3_Name objects = NULL;
            H3_Status status;
            uint32_t totalObjects = 0;
            uint32_t offset = 0;
            uint32_t nextOffset = 0;

            // List all the objects that have an ReadOnlyAfter metadata
            while ((status = H3_ListObjectsWithMetadata(h3_handle, &auth, current_bucket, "ReadOnlyAfter", offset, &objects, &totalObjects, &nextOffset)) == H3_CONTINUE || (status == H3_SUCCESS && totalObjects)) {
                H3_Name current_object;
                uint32_t current_object_pos = 0;
                size_t current_object_len = 0;
                int object;
                for (object = 0; object < totalObjects; object++) {
                    current_object = &(objects[current_object_pos]);
                    current_object_len = strlen(current_object);
                    current_object_pos += current_object_len;
                    while (objects[current_object_pos] == '\0')
                        current_object_pos++;

                    void* readOnly;
                    size_t readOnlySize;
                    // Read the ExpiresAt value
                    if (H3_ReadObjectMetadata(h3_handle, &auth, current_bucket, current_object, "ReadOnlyAfter", &readOnly, &readOnlySize) == H3_SUCCESS) {
                        double readOnlyTime = 0.0f;
                        H3_ObjectInfo objectInfo;
                        
                        // Copy the readOnly;
                        memcpy(&readOnlyTime, readOnly, readOnlySize);

                        if (H3_InfoObject(h3_handle, &auth, current_bucket, current_object, &objectInfo) == H3_SUCCESS) {
                            // Check if the object must become readonly
                            fprintf(stdout, "Now : [%lf] ~ ReadOnlyTime : [%lf]\n", TIMESPEC_TO_DOUBLE(now), TIMESPEC_TO_DOUBLE(objectInfo.lastModification) + readOnlyTime);
                            if (TIMESPEC_TO_DOUBLE(objectInfo.lastModification) + readOnlyTime >= TIMESPEC_TO_DOUBLE(now)) {
                                H3_Attribute readOnly;
                                readOnly.type = H3_ATTRIBUTE_READ_ONLY;
                                readOnly.readOnly = TRUE;
                                
                                H3_SetObjectAttributes(h3_handle, &auth, current_bucket, current_object, readOnly);
                            }
                        } 
                    }

                    // Free 
                    if (readOnly)
                        free(readOnly);
                }

                offset = nextOffset;
            }

            // Free 
            if (objects)
                free(objects);
        }
    }
 
    // Free resources
    if (buckets) 
        free(buckets);
    if (url)    
        free(url);
    H3_Free(h3_handle);

    return 0;
}